src/chainrules.jl

```diff
@@ -22,3 +22,33 @@ function ChainRulesCore.rrule(::Type{<:Base.Matrix}, B::T) where {T<:BlockDiagon
     return Matrix(B), Matrix_pullback
 end
 
+# multiplication
+function ChainRulesCore.rrule(::typeof(*), bm::BlockDiagonal{T, V}, v::AbstractVector{T}) where {T<:Union{Real, Complex}, V<:AbstractArray{T, 2}}
```

Comment:
The only change I would say that you should make is to restrict `V<:Matrix{T}` and `v::StridedVector{T}`. This rule won't do what we want if, say, `v` is a `FillArray` or the blocks of `bm` are themselves `Diagonal` or `BlockDiagonal`.